corehq/apps/domain/auth.py

```diff
@@ -19,53 +19,48 @@
 TOKEN = 'token'
 
 
-def determine_authtype_from_header(request, default=DIGEST):
+def determine_authtype_from_header(request, support_non_j2me_digest=True):
     """
     Guess the auth type, based on the headers found in the request.
+
+    CommCare mobile typically sends an unauthenticated request first, and
+    expects a basic auth challenge in response. This is at odds with the spec
+    for digest auth, in which the client sends an unauthenticated response and
+    receives a digest auth challenge. This means we can't support both CommCare
+    mobile and digest auth at the same endpoint.
+
+    For non-mobile endpoints (such as APIs), we can support basic, digest,
+    token, and apikey by defaulting to digest, since in all other cases, the
+    client should send the initial request with an Authorization header.
     """
     auth_header = (request.META.get('HTTP_AUTHORIZATION') or '').lower()
     if auth_header.startswith('basic '):
         return BASIC
-    elif auth_header.startswith('digest '):
-        return DIGEST
```

Comment:
I don't think I understand why this was removed. if the header is specifically asking for digest why would we not return it?

Reply:
+1 to @czue seems like we should keep it and document that so that if someone needs to use digest on an endpoint that defaults to basic auth they can

Reply:
I removed this because I thought it was misleading, but maybe it'd be better to keep it in there with a comment.  I'll add it back.  The first request in a digest auth transaction doesn't contain this header, so it's not a reliable way of determining that the client intends to use digest auth.  However, the second request _does_ include this header, so it seems useful to keep this here for that.

Cal, are you saying we should keep it in and ask clients to manually insert a header that matches this?  It does seem reasonable to require an explicit declaration of what type of auth they want to use, but I think we should do something separate, so as not to confuse the digest auth spec.  The `secure_post` endpoint accepts an `authtype` GET param - we could use that, or make a separate authtype header or something.

Reply:
I think it is more about giving them the option to insert a header like that, but generally yes. Either they learn to handle whatever auth we throw at them, something most code probably already can, since its likely using some http library that does auth transparently without burdening the programmer with protocol details, or if they require a certain type they can add that header, and we promise to support them. That is not an ideal world, but one that probably can make almost everyone happy.

Reply:
> The secure_post endpoint accepts an authtype GET param - we could use that, or make a separate authtype header or something.
I like the header "Authentication" that we already use most places better than the get param, because it seems more correct for reasons I cannot quite articulate. Get params seem like something that should affect the view/response, while headers are something that affect the http details like authentication type. In my mind you shouldn't get the same response from using opposite get params (like two different types of auth) but the header should not affect the response so it is reasonable to choose auth types that way.